### ISSUE TITLE
Authentication issue 

### ISSUE THREAD

**Issue by raviinder** [TakahikoKawasaki/nv-websocket-client#27]
Hi , 
    When i am trying to connect to my server which requires username/password.
 It respond me with 
**********************************************
HTTP Version  = HTTP/1.1
Status Code   = 401
Reason Phrase = Unauthorized
*************************************************

1) Do we have any kind of function by which we can add username/password to any initial connect to Websocket server ?

Here is my piece of code 
        try {
        	
		        HostnameVerifier hostnameVerifier = new HostnameVerifier() {
			    @Override
			    public boolean verify(String hostname, SSLSession session) {
			        return true;
			    }
			};
      	 TrustManager[] trustAllCerts = new TrustManager[]{
	               new X509TrustManager() {
	               public X509Certificate[] getAcceptedIssuers() {
	                         X509Certificate[] myTrustedAnchors = new X509Certificate[0];
		                    return myTrustedAnchors;
		              }
		            @Override
		              public void checkClientTrusted(X509Certificate[] certs, String authType) {
		               }
		             @Override
		               public void checkServerTrusted(X509Certificate[] certs, String authType) {
		                }
		           }
		        };
            
            // Create an SSLContext that uses our TrustManager
            
			try {
				sc = SSLContext.getInstance("SSL");
				 sc.init(null, trustAllCerts, null);
			} catch (NoSuchAlgorithmException | KeyManagementException e) {
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
           
            Authenticator.setDefault(new Authenticator(){
                protected PasswordAuthentication getPasswordAuthentication() {
                	String  username= "su";
                	String passsetN="cws";
                    return new PasswordAuthentication(username,passsetN.toCharArray());
                }});

        	
        	WebSocketFactory factory = new WebSocketFactory();
        	factory.setSSLSocketFactory(sc.getSocketFactory());
        	factory.createSocket("https://192.168.0.8/notification/echo.html")
			    .addListener(new WebSocketAdapter() {
			        @Override
			        public void onTextMessage(WebSocket ws, String message) {
			            // Received a response. Print the received message.
			            System.out.println(message);

			            // Close the WebSocket connection.
			            ws.disconnect();
			        }
			    })
			    .connect()
			    .sendText("Hello.");
		} catch ( IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
        catch (OpeningHandshakeException e)
        {
            // Status line.
            StatusLine sl = e.getStatusLine();
            System.out.println("=== Status Line ===");
            System.out.format("HTTP Version  = %s\n", sl.getHttpVersion());
            System.out.format("Status Code   = %d\n", sl.getStatusCode());
            System.out.format("Reason Phrase = %s\n", sl.getReasonPhrase());

            // HTTP headers.
            Map<String, java.util.List<String>> headers = e.getHeaders();
            System.out.println("=== HTTP Headers ===");
            for (Entry<String, java.util.List<String>> entry : headers.entrySet())
            {
                // Header name.
                String name = entry.getKey();

                // Values of the header.
                List values = (List) entry.getValue();

                //if (values == null || values.size() == 0)
      //          {
                    // Print the name only.
                    System.out.println(name);
                    continue;
                }

    //            for (String value : values)
                {
                    // Print the name and the value.
                //    System.out.format("%s: %s\n", name, value);
                }
            }

**************************************************************

**Comment by TakahikoKawasaki** [comment]
If your server requires [Basic Authentication](http://tools.ietf.org/html/rfc2617#section-2), in other words, if your server requries `Authorization` header whose content starts with `Basic`, it is easy to set your username and password. Use `WebSocket.setUserInfo(String, String)` method like below.

```java
String username = ...;
String password = ...;

ws.setUserInfo(username, password);
```

**Comment by raviinder** [comment]
Yup thank you it worked for me now but still while accessing my server it gives me OpeningHandshakeException  exception & when i print everything seems to ok to me .. 

Any clues ???
********************************************************************************************

=== Status Line ===
HTTP Version  = HTTP/1.1
Status Code   = 200
Reason Phrase = OK
=== HTTP Headers ===
Accept-Ranges: bytes
Connection: close
Content-Length: 1961
Content-Type: text/html
Date: Thu, 14 Jan 2016 01:20:50 GMT
ETag: "7a9-527082d225680"
Last-Modified: Wed, 16 Dec 2015 18:35:54 GMT
Server: Apache
Set-Cookie: wsvr_session=::ffff:10.128.34.91&su&1452734051&073984d9836e81779d20cacdb1a8b383261b712ce51e64c9e05564167508c305; Expires=Thu, 14-Jan-2016 01:24:11 GMT; path=/; httponly

**Comment by TakahikoKawasaki** [comment]
If the endpoint of your server can talk the WebSocket Protocol, it returns "101 Switching Protocols" as the HTTP status code. However, the output you got says "200 OK". This means that the endpoint you tried to connect to is not a WebSocket endpoint but a normal HTTP endpoint. Either the URL you specified or the configuration of the server is wrong.

**Comment by raviinder** [comment]
URL is good If i test URL on chrome or firefox browser it works perfectly that means URL is correct.

**Comment by raviinder** [comment]
So when i read document of my server. It says following 
"For  server connection  use the authentication token and HTTP cookie automatically generated for the current session in the response to a successful login request. The authentication token needs to be passed as a query parameter with parameter name -xsrf-."    Can we achieve this one using connect again back after getting 401 reply.

**Comment by TakahikoKawasaki** [comment]
Web browsers sometimes hide the scheme part (`http` or `https`). Check whether the scheme you passed to Web browsers and that to nv-websocket-client are equal.

**Comment by TakahikoKawasaki** [comment]
It sounds that the steps you should take are as follows.

1. Connect to the server using a normal HTTP client.
2. Receive a response from the server.
3. Extract an authentication token and a cookie from the response.
4. Connect to the server again using a WebSocket client with the authentication token and the cookie.

**Comment by raviinder** [comment]
Hi 
   If we want to enable debug mode on i.e. which headers my nv-websocket-client is sending. Can we enable some logs or print somewhere to see it.

**Comment by TakahikoKawasaki** [comment]
I added `onSendingHandshake` method to `WebSocketListener` for you and released a new version, `1.21`.

You can see the content of the opening handshake request from nv-websocket-client to your WebSocket server by implementing `onSendingHandshake` like below.

```java
public void onSendingHandshake(
    WebSocket websocket, String requestLine, List<String[]> headers) throws Exception
{
    // Print the request line, "GET {path} HTTP/1.1".
    System.out.println(requestLine);

    // For each header.
    for (String[] header : headers)
    {
        // Print the header, "{name}: {value}".
        System.out.format("%s: %s\n", header[0], header[1]);
    }
}
```

**Comment by raviinder** [comment]
Thank you .. that was really helpful. 
 But i have one more question do we have
